### ISSUE TITLE
netcdftime.datetime: implement __add__, __radd__, __sub__

### ISSUE THREAD

**Issue by ckhroulev** [Unidata/netcdf4-python#593]
I propose adding `__add__`, `__radd__`, `__sub__` implementations to `netcdftime.datetime`. This will make is possible to do things like

``` python
import netCDF4, datetime
print(netCDF4.netcdftime.datetime(-5000, 1, 1, 12) + datetime.timedelta(seconds=0))
```

This is prompted by #584 (the return value of `netcdftime.utime.num2date` should always work with `datetime.timedelta`).

Let me know if I should go ahead with this and submit a pull request.

The implementation I have in mind would use `utime.date2num`, shift the Julian day by `timedelta.total_seconds / 86400` and `utime.num2date` to get a date back. This would require extending `netcdftime.datetime` in `netcdftime.py` instead of using Cython (which would lead to a circular dependency).

By the way: why is `netcdftime.datetime` written in Cython?

**Comment by shoyer** [comment]
We wrote this type in Cython because it was an easy way to get significant performance improvements. 

I'm all for adding more functionality to the netcdftime.datetime class, but this should not come at the price of performance. Fortunately, it's quite easy to move Python code into Cython modules, so this shouldn't be an issue.

**Comment by davidhassell** [comment]
I'd also like to request that the `replace` method be implemented too. Thanks!

**Comment by ckhroulev** [comment]
@shoyer Hmm. Right. I would need to come up with a way to test for performance degradation then. 

Do you know of a good way to measure the performance of `netcdftime.datetime`?

Do you recall which `netcdftime.datetime` method benefitted from using Cython?

Would the time reported by `python tst_netcdftime.py` be an adequate measure, in your opinion?

@davidhassell Sure!

**Comment by shoyer** [comment]
@ckhroulev See https://github.com/Unidata/netcdf4-python/pull/303 for the original PR porting `netcdftime.datetime` to Cython and [here](https://github.com/Unidata/netcdf4-python/pull/303/files/7ad575e01c2addb945c8b1d278a441edeaac8c96#r20033706) for some benchmarks. It made date2num about 2x faster.

I'm still confused why you think "This would require extending netcdftime.datetime in netcdftime.py instead of using Cython (which would lead to a circular dependency)."

Instead, you could move date2num/num2date into the Cython module. The code would not even need to change, since Cython supports a strict superset of valid Python.

**Comment by ckhroulev** [comment]
I also noticed that `netcdftime.datetime` comparisons (unlike `datetime.datetime`) ignore differences smaller than 1 second:
```
>>> from netcdftime import datetime as A
>>> from datetime import datetime as B
>>> B(2000, 1, 1) == B(2000, 1, 1, microsecond=500)
False
>>> A(2000, 1, 1) == A(2000, 1, 1, microsecond=500)
True
```

This should probably be a separate issue, though.

**Comment by ckhroulev** [comment]
> Instead, you could move date2num/num2date into the Cython module. The code would not even need to change, since Cython supports a strict superset of valid Python.

Agreed.

**Comment by ckhroulev** [comment]
@shoyer Thanks for links to the original PR and the benchmarks!

**Comment by jswhit** [comment]
resolved with pull request #594
